File: web/frontend/src/lib/utils/openapi.ts
// Arenet - Homelab-friendly reverse proxy with integrated security
// Copyright (C) 2026  Ludovic Ramos
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see https://www.gnu.org/licenses/.

// v2.39 — helpers for the in-app OpenAPI viewer (/api-docs): list the
// operations by tag, resolve local $refs, build example payloads.
// Only what Arenet's own document uses (OpenAPI 3.1, local refs).

/* eslint-disable @typescript-eslint/no-explicit-any */
export type OpenAPIDoc = Record<string, any>;
export type Schema = Record<string, any>;

export const HTTP_METHODS = ['get', 'post', 'put', 'patch', 'delete'] as const;
export type HttpMethod = (typeof HTTP_METHODS)[number];

/** One operation of the document. */
export interface Operation {
	key: string;
	method: HttpMethod;
	path: string;
	/** Path as sent by the browser (servers override applied). */
	fullPath: string;
	tag: string;
	summary: string;
	description: string;
	role: string;
	parameters: Schema[];
	requestBody?: Schema;
	responses: Record<string, Schema>;
}

/** Follows a local "#/a/b" reference; returns the input when not a ref. */
export function resolve(doc: OpenAPIDoc, node: any): any {
	let cur = node;
	for (let guard = 0; cur && typeof cur === 'object' && typeof cur.$ref === 'string' && guard < 20; guard++) {
		const ref: string = cur.$ref;
		if (!ref.startsWith('#/')) return cur;
		cur = ref
			.slice(2)
			.split('/')
			.reduce<any>((acc, part) => (acc && typeof acc === 'object' ? acc[part] : undefined), doc);
	}
	return cur;
}

/** The name of a "#/components/schemas/X" reference, or "". */
export function refName(node: any): string {
	const ref: unknown = node?.$ref;
	return typeof ref === 'string' ? ref.split('/').pop() ?? '' : '';
}

/** Every operation, in tag order then path order. */
export function listOperations(doc: OpenAPIDoc): Operation[] {
	const base: string = doc?.servers?.[0]?.url ?? '';
	const tagOrder: string[] = (doc.tags ?? []).map((t: { name: string }) => t.name);
	const ops: Operation[] = [];
	for (const [path, rawItem] of Object.entries<any>(doc.paths ?? {})) {
		const item = resolve(doc, rawItem) ?? {};
		const prefix = item.servers?.[0]?.url === '/' ? '' : base;
		for (const method of HTTP_METHODS) {
			const op = item[method];
			if (!op) continue;
			ops.push({
				key: `${method} ${path}`,
				method,
				path,
				fullPath: prefix + path,
				tag: op.tags?.[0] ?? 'Other',
				summary: op.summary ?? '',
				description: op.description ?? '',
				role: op['x-arenet-role'] ?? '',
				parameters: [...(item.parameters ?? []), ...(op.parameters ?? [])].map((p: any) => resolve(doc, p)),
				requestBody: op.requestBody ? resolve(doc, op.requestBody) : undefined,
				responses: Object.fromEntries(
					Object.entries<any>(op.responses ?? {}).map(([code, r]) => [code, resolve(doc, r)])
				)
			});
		}
	}
	const rank = (tag: string) => {
		const i = tagOrder.indexOf(tag);
		return i < 0 ? tagOrder.length : i;
	};
	return ops.sort((a, b) => rank(a.tag) - rank(b.tag) || a.path.localeCompare(b.path) || a.method.localeCompare(b.method));
}

/** Groups operations by tag, keeping order. */
export function groupByTag(ops: Operation[]): { tag: string; ops: Operation[] }[] {
	const groups: { tag: string; ops: Operation[] }[] = [];
	for (const op of ops) {
		const last = groups[groups.length - 1];
		if (last && last.tag === op.tag) last.ops.push(op);
		else groups.push({ tag: op.tag, ops: [op] });
	}
	return groups;
}

/** Case-insensitive match on method, path, summary and tag. */
export function matches(op: Operation, query: string): boolean {
	const q = query.trim().toLowerCase();
	if (q === '') return true;
	return `${op.method} ${op.fullPath} ${op.summary} ${op.tag}`.toLowerCase().includes(q);
}

/** The JSON schema of a request body or response, if any. */
export function jsonSchema(doc: OpenAPIDoc, bodyOrResponse: Schema | undefined): Schema | undefined {
	const content = bodyOrResponse?.content;
	if (!content) return undefined;
	const media = content['application/json'] ?? Object.values<any>(content)[0];
	return media?.schema ? resolve(doc, media.schema) : undefined;
}

/** An explicit example of a request body / response, if the document has one. */
export function explicitExample(bodyOrResponse: Schema | undefined): unknown {
	const content = bodyOrResponse?.content;
	if (!content) return undefined;
	const media = content['application/json'] ?? Object.values<any>(content)[0];
	if (!media) return undefined;
	if (media.example !== undefined) return media.example;
	const first = media.examples && Object.values<any>(media.examples)[0];
	return first?.value;
}

/** The main type of a schema ("string", "array"…), ignoring "null". */
export function schemaType(schema: Schema | undefined): string {
	if (!schema) return '';
	const t = schema.type;
	if (Array.isArray(t)) return t.find((x: string) => x !== 'null') ?? t[0] ?? '';
	if (t) return t;
	if (schema.properties) return 'object';
	if (schema.items) return 'array';
	return '';
}

/** Builds an example value from a schema (explicit example first). */
export function sampleFromSchema(doc: OpenAPIDoc, node: any, depth = 0): unknown {
	const schema = resolve(doc, node);
	if (!schema || depth > 6) return null;
	if (schema.example !== undefined) return schema.example;
	if (Array.isArray(schema.examples) && schema.examples.length > 0) return schema.examples[0];
	if (schema.default !== undefined) return schema.default;
	if (Array.isArray(schema.enum) && schema.enum.length > 0) return schema.enum[0];
	for (const combo of ['oneOf', 'anyOf'] as const) {
		if (Array.isArray(schema[combo]) && schema[combo].length > 0) return sampleFromSchema(doc, schema[combo][0], depth + 1);
	}
	if (Array.isArray(schema.allOf)) {
		return Object.assign({}, ...schema.allOf.map((s: any) => sampleFromSchema(doc, s, depth + 1) ?? {}));
	}
	switch (schemaType(schema)) {
		case 'object': {
			const out: Record<string, unknown> = {};
			for (const [k, v] of Object.entries<any>(schema.properties ?? {})) {
				if (resolve(doc, v)?.readOnly) continue;
				out[k] = sampleFromSchema(doc, v, depth + 1);
			}
			return out;
		}
		case 'array':
			return [sampleFromSchema(doc, schema.items, depth + 1)];
		case 'integer':
		case 'number':
			return schema.minimum ?? 0;
		case 'boolean':
			return false;
		case 'string':
			return schema.format === 'date-time' ? '2026-01-01T00:00:00Z' : '';
	}
	return null;
}

/** Whether a method changes state (the "try it" asks for confirmation). */
export function isMutating(method: HttpMethod): boolean {
	return method !== 'get';
}

/** Fills {param} placeholders; missing values stay as-is. */
export function fillPath(path: string, values: Record<string, string>): string {
	return path.replace(/\{([^}]+)\}/g, (m, name: string) =>
		values[name] ? encodeURIComponent(values[name]) : m
	);
}

File: internal/api/openapi.go
// Arenet - Homelab-friendly reverse proxy with integrated security
// Copyright (C) 2026  The Arenet Authors
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see https://www.gnu.org/licenses/.

package api

import (
	"embed"
	"encoding/json"
	"fmt"
	"io/fs"
	"net/http"
	"sort"
	"strings"
	"sync"

	"gopkg.in/yaml.v3"
)

// openAPIFiles holds the hand-written OpenAPI document (v2.39): base.yaml
// plus one fragment per API domain. See
// docs/superpowers/specs/2026-09-22-openapi-design.md.
//
//go:embed openapi/*.yaml
var openAPIFiles embed.FS

// openAPIBase is the fragment holding info, security and shared
// components; every other file adds paths and schemas.
const openAPIBase = "openapi/base.yaml"

// mergedOpenAPI caches the merged document (the files are embedded, so
// it never changes at runtime).
var mergedOpenAPI = sync.OnceValues(func() (map[string]any, error) { return buildOpenAPI(openAPIFiles) })

// buildOpenAPI merges the base document with every fragment. A path, a
// schema, a response or a parameter defined twice is an error.
func buildOpenAPI(fsys fs.FS) (map[string]any, error) {
	doc, err := readOpenAPIFile(fsys, openAPIBase)
	if err != nil {
		return nil, err
	}
	names, err := fs.Glob(fsys, "openapi/*.yaml")
	if err != nil {
		return nil, fmt.Errorf("openapi: list fragments: %w", err)
	}
	sort.Strings(names)
	paths := childMap(doc, "paths")
	components := childMap(doc, "components")
	for _, name := range names {
		if name == openAPIBase {
			continue
		}
		frag, err := readOpenAPIFile(fsys, name)
		if err != nil {
			return nil, err
		}
		if err := mergeInto(paths, childMap(frag, "paths"), name, "path"); err != nil {
			return nil, err
		}
		fragComponents := childMap(frag, "components")
		for _, kind := range []string{"schemas", "responses", "parameters"} {
			if err := mergeInto(childMap(components, kind), childMap(fragComponents, kind), name, kind); err != nil {
				return nil, err
			}
		}
	}
	addOperationIDs(paths)
	return doc, nil
}

// openAPIMethodOrder lists the HTTP methods an OpenAPI path item holds.
var openAPIMethodOrder = []string{"get", "put", "post", "delete", "patch"}

// addOperationIDs gives every operation without one a stable
// operationId derived from its method and path ("getRoutesById",
// "postRoutesByIdWafTest") — client generators need them.
func addOperationIDs(paths map[string]any) {
	for p, raw := range paths {
		item, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		for _, m := range openAPIMethodOrder {
			op, ok := item[m].(map[string]any)
			if !ok {
				continue
			}
			if id, _ := op["operationId"].(string); id == "" {
				op["operationId"] = deriveOperationID(m, p)
			}
		}
	}
}

// deriveOperationID builds method + CamelCase path segments, path
// parameters as "By<Name>".
func deriveOperationID(method, path string) string {
	var b strings.Builder
	b.WriteString(method)
	for _, seg := range strings.Split(path, "/") {
		if seg == "" {
			continue
		}
		if strings.HasPrefix(seg, "{") && strings.HasSuffix(seg, "}") {
			b.WriteString("By")
			seg = strings.Trim(seg, "{}")
		}
		for _, word := range strings.FieldsFunc(seg, func(r rune) bool { return r == '-' || r == '_' || r == '.' }) {
			b.WriteString(strings.ToUpper(word[:1]) + word[1:])
		}
	}
	return b.String()
}

// readOpenAPIFile decodes one YAML file into JSON-compatible values.
func readOpenAPIFile(fsys fs.FS, name string) (map[string]any, error) {
	raw, err := fs.ReadFile(fsys, name)
	if err != nil {
		return nil, fmt.Errorf("openapi: read %s: %w", name, err)
	}
	var v any
	if err := yaml.Unmarshal(raw, &v); err != nil {
		return nil, fmt.Errorf("openapi: parse %s: %w", name, err)
	}
	m, ok := jsonCompatible(v).(map[string]any)
	if !ok {
		return nil, fmt.Errorf("openapi: %s is not a mapping", name)
	}
	return m, nil
}

// jsonCompatible turns YAML maps with non-string keys (status codes
// such as 200) into map[string]any.
func jsonCompatible(v any) any {
	switch t := v.(type) {
	case map[string]any:
		for k, e := range t {
			t[k] = jsonCompatible(e)
		}
		return t
	case map[any]any:
		out := make(map[string]any, len(t))
		for k, e := range t {
			out[fmt.Sprint(k)] = jsonCompatible(e)
		}
		return out
	case []any:
		for i, e := range t {
			t[i] = jsonCompatible(e)
		}
		return t
	}
	return v
}

// childMap returns m[key] as a map, creating it when absent.
func childMap(m map[string]any, key string) map[string]any {
	if c, ok := m[key].(map[string]any); ok {
		return c
	}
	c := map[string]any{}
	m[key] = c
	return c
}

// mergeInto copies src into dst, refusing keys already present.
func mergeInto(dst, src map[string]any, file, kind string) error {
	for k, v := range src {
		if _, dup := dst[k]; dup {
			return fmt.Errorf("openapi: %s %q defined twice (again in %s)", kind, k, file)
		}
		dst[k] = v
	}
	return nil
}

// openAPIJSON renders the merged document with info.version set to the
// running binary's version.
func openAPIJSON(version string) ([]byte, error) {
	doc, err := mergedOpenAPI()
	if err != nil {
		return nil, err
	}
	out := make(map[string]any, len(doc))
	for k, v := range doc {
		out[k] = v
	}
	info := map[string]any{}
	for k, v := range childMap(doc, "info") {
		info[k] = v
	}
	if version != "" {
		info["version"] = version
	}
	out["info"] = info
	return json.Marshal(out)
}

// getOpenAPI handles GET /api/v1/openapi.json.
func (h *Handler) getOpenAPI(w http.ResponseWriter, _ *http.Request) {
	body, err := openAPIJSON(h.version)
	if err != nil {
		h.logger.Error("openapi document", "err", err)
		writeError(w, http.StatusInternalServerError, "openapi document unavailable")
		return
	}
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(body)
}
